crates/typst-library/src/visualize/image/svg.rs

```diff
@@ -287,3 +311,124 @@ impl FontResolver<'_> {
+    svg_parent: Option<FileId>,
+    /// The first error that occurred when loading a linked image, if any.
+    error: Option<LoadError>,
+}
+
+impl<'a> ImageResolver<'a> {
+    fn new(world: Tracked<'a, dyn World + 'a>, svg_parent: Option<FileId>) -> Self {
+        Self { world, svg_parent, error: None }
+    }
+
+    /// Load a linked image or return None if a previous image caused an error,
+    /// or if the linked image failed to load.
+    /// Only the first error message is retained.
+    fn load(&mut self, href: &str) -> Option<usvg::ImageKind> {
+        if self.error.is_none() {
+            match self.load_or_error(href) {
+                Ok(image) => Some(image),
+                Err(err) => {
+                    self.error = Some(LoadError::new(
+                        ReportPos::None,
+                        eco_format!("failed to load linked image {} in SVG", href),
+                        err,
+                    ));
+                    None
+                }
+            }
+        } else {
+            None
+        }
+    }
+
+    /// Load a linked image or return an error message string.
+    fn load_or_error(&mut self, href: &str) -> Result<usvg::ImageKind, EcoString> {
+        // If the href starts with "file://", strip this prefix to construct an ordinary path.
+        let href = if let Some(stripped) = href.strip_prefix("file://") {
+            stripped
+        } else {
+            href
+        };
+
+        // Do not accept absolute hrefs. They would be parsed in typst in a way
+        // that is not compatible with their interpretation in the SVG standard.
+        if href.starts_with("/") {
+            return Err(EcoString::from("absolute paths are not allowed"));
+        }
+
+        // Exit early if the href is an URL.
+        if let Some(pos) = href.find("://") {
+            let scheme = &href[..pos];
+            if scheme
+                .chars()
+                .all(|c| c.is_ascii_alphanumeric() || c == '+' || c == '-' || c == '.')
+            {
+                return Err(EcoString::from("URLs are not allowed"));
+            }
+        }
+
+        // Resolve the path to the linked image.
+        if self.svg_parent.is_none() {
+            return Err(EcoString::from("cannot access file system from here"));
+        }
+        // Replace the final underscore in svg_parent by the href.
+        let href_file = self.svg_parent.unwrap().join(href);
+
+        // Load image if file can be accessed.
+        match self.world.file(href_file) {
+            Ok(bytes) => {
+                let arc_data = Arc::new(bytes.to_vec());
+                let format = match determine_format_from_path(href) {
+                    Some(format) => Some(format),
+                    None => ImageFormat::detect(&arc_data),
+                };
+                match format {
+                    None => Err(EcoString::from("could not determine image format")),
```

Comment:
```suggestion
                    None => Err(EcoString::from("unknown image format")),
```
for consistency with the other code